File: apps/ios/ExpCore/Sources/Domain/ExpToolDisplay.swift
import Foundation

/// EXP-846 — how an **Exponential MCP tool call** renders in a transcript.
///
/// The agent calls our own MCP surface constantly (`exponential_issues_create`,
/// `exponential_pr_open`, …) and a row saying
/// `mcp__exponential__exponential_issues_create` told a reader nothing. The
/// contract's `expToolDisplay` tables name, per tool, the PROGRESSIVE caption
/// ("Creating issue"), the DONE caption ("Created issue"), which input field is
/// the subject, and what KIND of thing the answer is — so every client draws
/// the same row off the same strings.
///
/// Hand-mirrored ×4 (web `lib/agent-feed.ts`, Android `domain/ExpToolDisplay.kt`,
/// desktop `steer::exp_tool_display`); the tables themselves are GENERATED
/// (`DomainContract.expTool*`), so nothing here hard-codes a caption.
public struct ExpToolDisplay: Equatable, Sendable {
    /// The contract `expToolNames` row (`issues_create`), namespace stripped.
    public let row: String
    /// The caption while the call is in flight ("Creating issue").
    public let progressive: String
    /// The caption once it settled ("Created issue").
    public let done: String
    /// The INPUT field that names the subject (`title`, `id`, …); empty when
    /// the tool takes no subject (a list).
    public let subjectKey: String
    /// What the answer IS — what the row previews on completion.
    public let result: ExpToolResultKind

    public init(
        row: String,
        progressive: String,
        done: String,
        subjectKey: String,
        result: ExpToolResultKind
    ) {
        self.row = row
        self.progressive = progressive
        self.done = done
        self.subjectKey = subjectKey
        self.result = result
    }

    /// The caption a row wears: progressive while running, done once settled.
    public func caption(settled: Bool) -> String { settled ? done : progressive }

    /// The Exponential tool a call NAMES, or nil for every other tool.
    ///
    /// A name matches when it ends with a contract row AND the part in front of
    /// it ends with the contract prefix — so `exponential_issues_create`,
    /// `mcp__exponential__exponential_issues_create` and
    /// `exponential.exponential_issues_create` all resolve, while another
    /// server's `issues_create` (no `exponential_` in front) never does.
    /// Mirrors desktop `mapper::exp_tool_row`.
    public static func resolve(toolName: String) -> ExpToolDisplay? {
        let name = toolName.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !name.isEmpty else { return nil }
        let rows = DomainContract.expToolNames
        guard let index = rows.firstIndex(where: { row in
            guard !row.isEmpty, name.count > row.count, name.hasSuffix(row) else { return false }
            return name.dropLast(row.count).hasSuffix(DomainContract.expToolPrefix)
        }) else { return nil }
        // The five tables are generated from ONE contract list, so they are the
        // same length — the guard is a belt against a half-regenerated client,
        // which must render a plain tool row rather than crash.
        guard DomainContract.expToolProgressive.indices.contains(index),
              DomainContract.expToolDone.indices.contains(index),
              DomainContract.expToolSubjectKeys.indices.contains(index),
              DomainContract.expToolResults.indices.contains(index)
        else { return nil }
        return ExpToolDisplay(
            row: rows[index],
            progressive: DomainContract.expToolProgressive[index],
            done: DomainContract.expToolDone[index],
            subjectKey: DomainContract.expToolSubjectKeys[index],
            result: ExpToolResultKind(wire: DomainContract.expToolResults[index])
        )
    }
}

/// EXP-846: the contract `expToolDisplay.resultKinds` — what an Exponential
/// tool's answer IS, and therefore what its row previews. An unknown value
/// (a newer contract against an older client) reads as `.none`: a caption
/// without a preview, never a broken row.
public enum ExpToolResultKind: String, Sendable, CaseIterable {
    case none
    case issue
    case pr
    case comment
    case session
    case board
    case action
    case automation
    case list

    public init(wire: String) {
        // Spelled out rather than `?? .none`: a `none` CASE beside Optional's
        // own `none` is exactly where Swift's inference gets interesting.
        if let kind = ExpToolResultKind(rawValue: wire) {
            self = kind
        } else {
            self = ExpToolResultKind.none
        }
    }
}

File: apps/android/app/src/main/java/com/exponential/app/domain/ExpToolDisplay.kt
package com.exponential.app.domain

/**
 * EXP-846: how a transcript renders a tool call that is one of OUR OWN MCP
 * tools. An agent working on an issue spends half its calls talking back to
 * Exponential, and those rows read as `mcp__exponential__exponential_issues_create`
 * — the raw wire name — beside the file edits. They get the product's own
 * vocabulary instead: the Exponential mark, a caption in the tense the call is
 * in ("Creating issue" → "Created issue"), the SUBJECT the call named, and,
 * once it settles, a small preview of what came back ([ToolResultPreview]).
 *
 * Every string comes from the contract's `expToolDisplay` tables
 * (`DomainContract.expTool*`, generated ×4) — nothing here is written down
 * twice, so adding an MCP tool is a contract edit and a regenerate.
 */
data class ExpToolRow(
    /** The contract row this call matched (`issues_create`). */
    val row: String,
    /** The caption to render: the progressive form while the call is in
     *  flight, the done form once it settled. */
    val caption: String,
    /** Which INPUT field names the call's subject — `` when the call has no
     *  meaningful one (a plain list). */
    val subjectKey: String,
    /** What the answer is: a contract `expToolDisplay.resultKinds` value,
     *  deciding which preview (if any) the settled row draws. */
    val result: String,
)

object ExpToolDisplay {
    // The result kinds this build renders. An unknown kind (a newer contract)
    // falls back to caption-only, which is the [RESULT_NONE] behaviour.
    const val RESULT_NONE = "none"
    const val RESULT_ISSUE = "issue"
    const val RESULT_PR = "pr"
    const val RESULT_COMMENT = "comment"
    const val RESULT_SESSION = "session"
    const val RESULT_BOARD = "board"
    const val RESULT_ACTION = "action"
    const val RESULT_AUTOMATION = "automation"
    const val RESULT_LIST = "list"

    /**
     * The Exponential MCP tool a call NAMES, or null for anything else.
     * Byte-identical with the engine's `exp_tool_row` (mapper.rs) and its
     * siblings ×4: a wire name ENDING in `exponential_<row>` for a contract
     * `expToolNames` row, whatever namespace an adapter put in front of it
     * (`mcp__exponential__exponential_issues_create` on claude, the bare
     * `exponential_pr_open` elsewhere). The contract PREFIX is required —
     * that is what keeps another MCP server's `issues_create` out.
     */
    fun row(name: String?): String? {
        val trimmed = name?.trim().orEmpty()
        if (trimmed.isEmpty()) return null
        return DomainContract.expToolNames.firstOrNull { row ->
            trimmed.length > row.length &&
                trimmed.endsWith(row) &&
                trimmed.dropLast(row.length).endsWith(DomainContract.expToolPrefix)
        }
    }

    /**
     * The row's whole presentation, or null when the call is not one of ours.
     * [settled] is the call's own state — a status landed, so the caption flips
     * from the progressive form to the done one. A FAILED call keeps the
     * generic failed styling; the caption still reads in the past tense,
     * because the call is over either way.
     */
    fun forName(name: String?, settled: Boolean): ExpToolRow? {
        val row = row(name) ?: return null
        val at = DomainContract.expToolNames.indexOf(row)
        val captions = if (settled) DomainContract.expToolDone else DomainContract.expToolProgressive
        return ExpToolRow(
            row = row,
            caption = captions.getOrNull(at) ?: row,
            subjectKey = DomainContract.expToolSubjectKeys.getOrNull(at).orEmpty(),
            result = DomainContract.expToolResults.getOrNull(at) ?: RESULT_NONE,
        )
    }
}
